arch/xtensa/src/esp32s3/esp32s3_sdmmc.c

```diff
@@ -0,0 +1,2924 @@
+ *
+ * Returned Value:
+ *   OK is success; a negated errno on failure
+ *
+ ****************************************************************************/
+
+static int esp32s3_waitresponse(struct sdio_dev_s *dev, uint32_t cmd)
+{
+  volatile int32_t timeout;
+  clock_t watchtime;
+  uint32_t events;
+
+  mcinfo("cmd=%04x\n", cmd);
+
+  switch (cmd & MMCSD_RESPONSE_MASK)
+    {
+    case MMCSD_NO_RESPONSE:
+      events  = SDCARD_CMDDONE_STA;
+      timeout = SDCARD_CMDTIMEOUT;
+      break;
+
+    case MMCSD_R1_RESPONSE:
+    case MMCSD_R1B_RESPONSE:
+    case MMCSD_R2_RESPONSE:
+    case MMCSD_R4_RESPONSE:
+    case MMCSD_R5_RESPONSE:
+    case MMCSD_R6_RESPONSE:
+      events  = (SDCARD_CMDDONE_STA | SDCARD_RESPDONE_STA);
+      timeout = SDCARD_LONGTIMEOUT;
+      break;
+
+    case MMCSD_R3_RESPONSE:
+    case MMCSD_R7_RESPONSE:
+      events  = (SDCARD_CMDDONE_STA | SDCARD_RESPDONE_STA);
+      timeout = SDCARD_CMDTIMEOUT;
+      break;
+
+    default:
+      return -EINVAL;
+    }
+
+  /* Then wait for the response (or timeout or error) */
+
+  watchtime = clock_systime_ticks();
+  while ((esp32s3_getreg(ESP32S3_SDMMC_RINTSTS) & events) != events)
+    {
+      if (clock_systime_ticks() - watchtime > timeout)
+        {
+          mcerr("ERROR: Timeout cmd: %04x events: %04x STA: %08x "
+                "RINTSTS: %08x\n",
+                cmd, events, esp32s3_getreg(ESP32S3_SDMMC_STATUS),
+                esp32s3_getreg(ESP32S3_SDMMC_RINTSTS));
+
+          return -ETIMEDOUT;
+        }
+      else if ((esp32s3_getreg(ESP32S3_SDMMC_RINTSTS) & SDCARD_INT_RESPERR)
+               != 0)
+        {
+          mcerr("ERROR: SDMMC failure cmd: %04x events: %04x STA: %08x "
+                "RINTSTS: %08x\n",
+                cmd, events, esp32s3_getreg(ESP32S3_SDMMC_STATUS),
+                esp32s3_getreg(ESP32S3_SDMMC_RINTSTS));
+
+          return -EIO;
+        }
+    }
+
+  esp32s3_putreg(SDCARD_CMDDONE_CLEAR, ESP32S3_SDMMC_RINTSTS);
+  return OK;
+}
+
+/****************************************************************************
+ * Name: esp32s3_recv*
+ *
+ * Description:
+ *   Receive response to SDIO command.  Only the critical payload is
+ *   returned -- that is 32 bits for 48 bit status and 128 bits for 136 bit
+ *   status.  The driver implementation should verify the correctness of
+ *   the remaining, non-returned bits (CRCs, CMD index, etc.).
+ *
+ * Input Parameters:
+ *   dev    - An instance of the SDIO device interface
+ *   Rx - Buffer in which to receive the response
+ *
+ * Returned Value:
+ *   Number of bytes sent on success; a negated errno on failure.  Here a
+ *   failure means only a faiure to obtain the requested response (due to
+ *   transport problem -- timeout, CRC, etc.).  The implementation only
+ *   assures that the response is returned intacta and does not check errors
+ *   within the response itself.
+ *
+ ****************************************************************************/
```

Comment:
If possible, please provide a unique function description for each one of the following functions.